ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1098 @@
+			*value = &p_tlv[RPROC_TLV_VALUE_OF];
+			*length = tlv_length;
+			if (tlv_length)
+				return TEE_SUCCESS;
+			else
+				return TEE_ERROR_NO_DATA;
+		}
+		p_tlv += ROUNDUP_64(sizeof(struct remoteproc_tlv) + tlv_length);
+	}
+
+	return TEE_ERROR_NO_DATA;
+}
+
+static struct remoteproc_context *remoteproc_find_firmware(uint32_t rproc_id)
+{
+	struct remoteproc_context *ctx = NULL;
+
+	TAILQ_FOREACH(ctx, &firmware_head, link)
+		if (ctx->rproc_id == rproc_id)
+			return ctx;
+
+	return NULL;
+}
+
+static struct remoteproc_context *remoteproc_add_firmware(uint32_t rproc_id)
+{
+	struct remoteproc_context *ctx = NULL;
+
+	ctx = TEE_Malloc(sizeof(*ctx), TEE_MALLOC_FILL_ZERO);
+	if (!ctx)
+		return NULL;
+
+	ctx->rproc_id = rproc_id;
+
+	TAILQ_INSERT_TAIL(&firmware_head, ctx, link);
+
+	return ctx;
+}
+
+static const struct remoteproc_sig_algo *remoteproc_get_algo(uint32_t sign_type)
+{
+	unsigned int i = 0;
+
+	for (i = 0; i < ARRAY_SIZE(rproc_ta_sign_algo); i++)
+		if (sign_type == rproc_ta_sign_algo[i].sign_type)
+			return &rproc_ta_sign_algo[i];
+
+	return NULL;
+}
+
+static TEE_Result remoteproc_pta_verify(struct remoteproc_context *ctx,
+					const struct remoteproc_sig_algo *algo,
+					uint8_t *hash, uint32_t hash_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct remoteproc_fw_hdr *hdr = (void *)ctx->sec_cpy;
+	struct rproc_pta_key_info *keyinfo = NULL;
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_INPUT);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	size_t length = 0;
+	uint8_t *tlv_keyinfo = NULL;
+	uint8_t *sign = NULL;
+
+	res = remoteproc_get_tlv(ctx->tlvs, hdr->tlv_len,
+				 RPROC_TLV_PKEYINFO, &tlv_keyinfo,
+				 &length);
+	if (res != TEE_SUCCESS && res != TEE_ERROR_NO_DATA)
+		return res;
+
+	keyinfo = TEE_Malloc(sizeof(*keyinfo) + length, TEE_MALLOC_FILL_ZERO);
+	if (!keyinfo)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	keyinfo->algo = algo->id;
+	keyinfo->info_size = length;
+	memcpy(keyinfo->info, tlv_keyinfo, length);
+
+	sign = FW_SIGN_PTR(ctx->sec_cpy, hdr);
+
+	params[0].value.a = ctx->rproc_id;
+	params[1].memref.buffer = keyinfo;
+	params[1].memref.size = rproc_pta_keyinfo_size(keyinfo);
+	params[2].memref.buffer = hash;
+	params[2].memref.size = hash_len;
+	params[3].memref.buffer = sign;
+	params[3].memref.size = hdr->sign_len;
+
+	res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+				  PTA_RPROC_VERIFY_DIGEST,
+				  param_types, params, NULL);
+	if (res != TEE_SUCCESS)
+		EMSG("Failed to verify signature, res = %#"PRIx32, res);
+
+	TEE_Free(keyinfo);
+
+	return res;
+}
+
+static TEE_Result
+remoteproc_save_fw_header_and_tlvs(struct remoteproc_context *ctx,
+				   void *fw_orig, uint32_t fw_orig_size)
+{
+	struct remoteproc_fw_hdr *hdr = fw_orig;
+	uint32_t length = 0;
+
+	if (ADD_OVERFLOW(sizeof(*hdr), ROUNDUP_64(hdr->tlv_len), &length) ||
```

Comment:
Thi should check `hdr` alignment `!IS_ALIGNED_WIT(fw_orig, struct remoteproc_fw_hdr))`.